lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -676,6 +676,104 @@ class DecomposeAtenTOp : public OpRewritePattern<AtenTOp> {
 };
 } // namespace
 
+// Decompose aten.repeat into aten.expand and aten.view ops.
+//
+// Ref: https://pytorch.org/docs/stable/generated/torch.Tensor.repeat.html
+//
+// For shape [S1, S2, S3] and repeats [M0, M1, M2, M3]
+//     MS0 = M0; MS1 = M1 * S1; MS2 = M2 * S2; MS3 = M3 * S3
+//
+// def aten_repeat(self, repeats):
+//     sizes = self.size()
+//     new_sizes0 = []
+//     new_sizes1 = []
+//     new_sizes2 = []
+//     leading_rank = repeats.size() - sizes.size()
+//     for r in range(leading_rank):
+//         new_sizes0.append(1)
+//         new_sizes1.append(repeats[r])
+//         new_sizes2.append(repeats[r])
+//
+//     for s, m in zip(sizes, repeats[leading_rank:]):
+//         n_sizes0 += [1, s]
+//         n_sizes1 += [m, s]
+//         n_sizes2 += [m * s]
+//     return self.view(new_sizes0).expand(new_sizes1).view(n_sizes2)
+//
+namespace {
+class DecomposeAtenRepeatOp : public OpRewritePattern<AtenRepeatOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenRepeatOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value self = op.self();
+    MLIRContext *context = op.getContext();
+    auto rank = getTensorRank(self);
+    if (rank < 0)
+      return rewriter.notifyMatchFailure(op, "Unimplemented: unranked tensor");
+
+    SmallVector<Value> repeats;
+    if (!getListConstructElements(op.repeats(), repeats))
+      return rewriter.notifyMatchFailure(
+          op, "Unimplemented: repeats not list of Scalar");
+
+    if (rank > repeats.size()) {
+      return rewriter.notifyMatchFailure(
+          op, "repeats are not matched with self's rank");
+    }
+
+    SmallVector<Value> newSizes0, newSizes1, newSizes2;
+    Value one = rewriter.create<Torch::ConstantIntOp>(
+        loc, rewriter.getI64IntegerAttr(1));
+
+    // leadingRank >= 0
+    auto leadingRank = repeats.size() - rank;
+    for (size_t i = 0; i < leadingRank; ++i) {
+      newSizes0.push_back(one);
+      newSizes1.push_back(repeats[i]);
+      newSizes2.push_back(repeats[i]);
+    }
+    for (size_t i = 0; i < rank; i++) {
+      Value dim = rewriter.create<Torch::ConstantIntOp>(
+          loc, rewriter.getI64IntegerAttr(i));
+      auto dimSize = rewriter.create<AtenSizeIntOp>(loc, self, dim);
+      auto scale = repeats[i + leadingRank];
+      newSizes0.insert(newSizes0.end(), {one, dimSize});
+      newSizes1.insert(newSizes1.end(), {scale, dimSize});
+      Value scaledSize = rewriter.create<AtenMulIntOp>(loc, dimSize, scale);
+      newSizes2.push_back(scaledSize);
+    }
+
+    auto newRank = repeats.size();
+    SmallVector<int64_t> newShape0(rank + newRank, ShapedType::kDynamicSize);
```

Comment:
Let's not make it dynamic when we can infer the shape. Otherwise, let's make it dynamic.

Reply:
I have tried static case it also raised an exception:
```
Exception:
        Lower Torch Backend IR -> Linalg-on-Tensors Backend IR failed with the following diagnostics:
        error: failed to legalize operation 'torch.aten.view' that was explicitly marked illegal
        note: see current operation: %115 = "torch.aten.view"(%114, %70) : (!torch.vtensor<[2,1,3,3,2,4,1],f32>, !torch.list<int>) -> !torch.vtensor<[2,3,6,4],f32>                                                                                                                       


        Error can be reproduced with:
        $ torch-mlir-opt -pass-pipeline='torch-backend-to-linalg-on-tensors-backend-pipeline' /tmp/RepeatModule.mlir
        Add '-mlir-print-ir-after-all -mlir-disable-threading' to get the IR dump for debugging purpose.
```

Reply:
The TorchToLinalg seems just capable to handle simple uses of view for now.

Reply:
Oh! The problem is with collapsing right?

Reply:
Does the expansion work?

Reply:
Yes, I think it's the problem of collapsing.